[Draft] Add slot

alexfauquette opened the pull request:
In his PR a component `PickersActionBar` is created

To let typescript pass, I did the following:
- remove the unsuported props (`clearText`, `cancelText`, `showTodayButton`, ...) (easy part)
- updated the type of `components` and `componentSlot` (boring part)

For the second point, it is done as follow
```ts
// Export the interface of slot and slot props to be able to reuse them later
export interface MobileWrapperSlotsComponent extends PickersModalDialogSlotsComponent, DateInputSlotsComponent { }
export interface MobileWrapperSlotsComponentProps extends PickersModalDialogSlotsComponentsProps { }


export interface InternalMobileWrapperProps extends MobileWrapperProps, PickerStateWrapperProps {
  DateInputProps: DateInputPropsLike & { ref?: React.Ref<HTMLDivElement> };
  PureDateInputComponent: React.JSXElementConstructor<DateInputPropsLike>;

  // Set the new interface in the props interface
  components?: Partial<MobileWrapperSlotsComponent>;
  componentsProps?: Partial<MobileWrapperSlotsComponentProps>;
}
```

This is done for the two wrappers (desktop and mobile) which get slots from the modal/dialog, but also `PickerStateWrapperProps`

It's also done for all the other wrapped components: `<[Mobile|Desktop][Time|Date|DateTime]Piker />` which get slots from the wrapper, and the picker


## Todo

- [ ] test the action bar
- [ ] correct tests that were using removed props

flaviendelangle commented on packages/x-date-pickers/src/DesktopDateTimePicker/DesktopDateTimePicker.tsx:
```diff
@@ -4,20 +4,35 @@ import {
 import { useDateTimeValidation } from '../internals/hooks/validation/useDateTimeValidation';
 import { KeyboardDateInput } from '../internals/components/KeyboardDateInput';
 import { usePickerState } from '../internals/hooks/usePickerState';
+import { CalendarOrClockPickerSlotsComponentsProps } from '../internals/components/CalendarOrClockPicker/CalendarOrClockPicker';
```
Could be re-exported from `CalendarOrClockPicker/index.ts` (x other similar components)

flaviendelangle commented on packages/x-date-pickers/src/DesktopDateTimePicker/DesktopDateTimePicker.tsx:
```diff
@@ -71,15 +86,17 @@ export const DesktopDateTimePicker = React.forwardRef(function DesktopDateTimePi
       TransitionComponent={TransitionComponent}
-      clearable={clearable}
-      clearText={clearText}
+      components={components}
```
I wonder if it would be safer to just keep `components` and `componentsProps` in `other` and do `components={other.components}` here (same for `componentsProps`).

Same for any similar scenario where we remove it from the spread props because we need to both pass it to a component receiving the spread props and to a component not receiving them.

I fixed at least two bugs on the picker where a prop was removed from `other` to be passed to a component but where the dev forgot to also pass it explicitly to the component receiving `{...other}`.

alexfauquette replied:
I'm not sure it will be safer. The danger to remove one props is during this PR (I promise I'm careful) 

After this modification, to "accidentally" stop passing prop `component` you need to delete the line `component={component}`

Whereas with `components={other.components}` you don't know if it intentional that `components` is still in `other`

flaviendelangle replied:
I was not thinking just for this PR but in general.
But I'm not sure it's a good idea, just those spread in all direction are quite a mess :grimacing:

flaviendelangle commented on packages/x-date-pickers/src/internals/components/PickersModalDialog.tsx:
```diff
@@ -1,48 +1,18 @@
+import { PickersActionBarProps } from '../../PickersActionBar';
+
+export interface PickersModalDialogSlotsComponent {
+  ActionBar: React.ElementType;
```
```suggestion
  ActionBar: React.ElementType<PickersActionBarProps>;
```

flaviendelangle commented on packages/x-date-pickers/src/internals/components/PickersPopper.tsx:
```diff
@@ -334,13 +309,16 @@ export const PickersPopper = (props: PickerPopperProps) => {
+                  onClear={onClear}
+                  onCancel={onCancel}
+                  onSetToday={onSetToday}
+                  actions={['clear']}
```
We will have to be careful here
If the user sets `props.componentsProps.actionBar.onCancel` for instance, I would expect it to call both our callback passed from `onPickerState` AND it's custom callback (like we do on events like `onClick` on the grid).

We could create a small wrapper handling that
Or at least remove those 4 `onXXX` prop from the typing to avoid the temptation of overriding them.

alexfauquette replied:
I would expect the same, but for now, I simply remove them from the type with `Omit<PickersActionBarProps, 'onAccept' | 'onClear' | 'onCancel' | 'onSetToday'>;` such that it is clearly visible which prop are not handled yet

Let see first how this slot will be used

flaviendelangle replied:
With https://github.com/mui/mui-x/pull/4758 you can do `Omit<PickerActionBarProps, keyof PickerStateProps>`

flaviendelangle reviewed (commented):
It is what I have in mind yes

The TS definition is very verbose with all those components but I think it's best to remain explicit here.

alexfauquette committed:
[pickers] Add slot for actionBar

make typescript pass on comunity version

prettier

fix props propagation

simplify typescript exportation

fix typescript

fix problems with defaults

fix tests

expose slots for consistency

scripts

add tests on PickersActionBar

renovate[bot] committed:
Bump typescript-eslint to ^5.24.0 (#4801)

Co-authored-by: Renovate Bot <bot@renovateapp.com>

renovate[bot] committed:
Bump @types/react to ^17.0.45 (#4797)

Co-authored-by: Renovate Bot <bot@renovateapp.com>

alexfauquette committed:
[pickers] Add l10n support (#4517)

alexfauquette committed:
Merge remote-tracking branch 'upstream/master' into add-slots

alexfauquette committed:
fix tests

alexfauquette committed:
fix types

alexfauquette committed:
proptypes & docs:api

alexfauquette committed:
fix plurial on componentsProps

alexfauquette committed:
scripts

alexfauquette committed:
use call back to modify the actions depending to the context

alexfauquette committed:
add tests for callback
